src/middlewared/middlewared/plugins/rsync.py

```diff
@@ -0,0 +1,252 @@
+                        try:
+                            progress = int([x for x in message.split(' ') if '%' in x][0][:-1])
+                        except (IndexError, ValueError):
+                            pass
+                except BaseException as err:
+                    # Catch IOERROR Errno 9 which usually arises because
+                    # of already closed fileobject being used here therby
+                    # raising Bad File Descriptor error. In this case break
+                    # and the outer while loop will check for rsync_proc.poll()
+                    # to be None or not and DTRT
+                    if hasattr(err, 'errno') and err.errno == 9:
+                        break
+                    logger.debug('Error whilst parsing rsync progress', exc_info=True)
+
+        except BaseException as e:
+            raise CallError(f'Rsync copy job id: {job.id} failed due to: {e}', errno.EIO)
+
+        if rsync_proc.returncode != 0:
+            job.set_progress(None, 'Rsync copy job failed')
+            raise CallError(
+                f'Rsync copy job id: {job.id} returned non-zero exit code. Command used was: {line}. Error: {rsync_proc.stderr.read()}'
+            )
+
+    @accepts(Dict(
+        'rsync-copy',
+        Str('user', required=True),
+        Str('path', required=True),
+        Str('remote_user'),
+        Str('remote_host', required=True),
+        Str('remote_path'),
+        Int('remote_ssh_port'),
+        Str('remote_module'),
+        Str('direction', enum=['PUSH', 'PULL'], required=True),
+        Str('mode', enum=['MODULE', 'SSH'], required=True),
+        Str('remote_password'),
+        Dict(
+            'rsync-properties',
+            Bool('recursive'),
+            Bool('compress'),
+            Bool('times'),
+            Bool('archive'),
+            Bool('delete'),
+            Bool('preserve_permissions'),
+            Bool('preserve_attributes'),
+            Bool('delay_updates')
+        ),
+        required=True
+    ))
+    @job()
+    def copy(self, job, rcopy):
+        """
+        Starts an rsync copy task between current freenas machine
+        and specified remote host (or local copy too). It reports
+        the progress of the copy task.
+        """
+
+        # Assigning variables and such
+        user = rcopy.get('user')
+        path = rcopy.get('path')
+        mode = rcopy.get('mode')
+        remote_path = rcopy.get('remote_path')
+        remote_host = rcopy.get('remote_host')
+        remote_module = rcopy.get('remote_module')
+        remote_user = rcopy.get('remote_user', rcopy.get('user'))
+        remote_address = remote_host if '@' in remote_host else f'"{remote_user}"@{remote_host}'
+        remote_password = rcopy.get('remote_password', None)
+        password_file = None
+        properties = rcopy.get('rsync-properties', defaultdict(bool))
+
+        # Let's do a brief check of all the user provided parameters
+        if is_empty(path):
```

Comment:
Using is_empty() seems superfluous, empty string nad None is caught by `if path:`, I am willing to ignore a bunch of spaces